### PR TITLE
[IPU] support dy2static for IPU merge code

### PR DESCRIPTION
**Pull request by gglin001**
### PR types

Others
### PR changes

Others
### Describe

clone of https://github.com/PaddlePaddle/Paddle/pull/42952

~fix pre-commit CI~

文档 PR https://github.com/PaddlePaddle/docs/pull/4932

文档预览链接：http://10.136.157.23:8090/documentation/docs/zh/faq/index_cn.html?reviewVersion=jenkins-doc-review-2374

image

### CODE REVIEW

**Review by qili93 — DISMISSED** [review]
LGTM, code review in https://github.com/PaddlePaddle/Paddle/pull/42952

File: python/paddle/fluid/layers/loss.py [hunk 1]
```diff
@@ -1230,6 +1231,54 @@ def softmax_with_cross_entropy(logits,
         return_softmax, axis)
 
 
+def identity_loss(x, reduction="none"):
+    r"""Marks a tensor as being part of the loss calculation for IPU.
+
+    This function should be called on the (final) loss of a model so that
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/loss.py]: ~~This function~~

File: python/paddle/fluid/layers/loss.py [hunk 2]
```diff
@@ -1230,6 +1231,54 @@ def softmax_with_cross_entropy(logits,
...
+    it is used as the start of backpropagation. This is equivalent to calling
+    ``x.backward()`` on a tensor ``x`` when running on the CPU.
+
+    Parameters:
+        x (Variable): The calculated loss ``Tensor``.
+        reduction(str|int): Reduce the loss output. The default value is "none". Supported values are:
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/loss.py]: reduction(str|int, optional)

  请添加 reduction 为 int 时的说明。

Comment (gglin001) [reply python/paddle/fluid/layers/loss.py]: fixed,

```
            * ``"sum"``: Sum the losses.
```

to 

```
            * ``"sum"`` or ``1``: Sum the losses.
```

File: python/paddle/fluid/layers/loss.py [hunk 3]
```diff
@@ -1230,6 +1231,54 @@ def softmax_with_cross_entropy(logits,
         return_softmax, axis)
 
 
+def identity_loss(x, reduction="none"):
+    r"""Marks a tensor as being part of the loss calculation for IPU.
+
+    This function should be called on the (final) loss of a model so that
+    it is used as the start of backpropagation. This is equivalent to calling
+    ``x.backward()`` on a tensor ``x`` when running on the CPU.
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/loss.py]: This is equivalent to calling ``x.backward()`` on a tensor ``x`` when running on the CPU. 中文文档没有；
此外，关于 reduction ，英文版在参数中进行说明，而中文在描述中与参数中都说明了；建议按中文版，在功能描述中说明 reduction 为 string/int 时的行为，在参数描述中，保留 指定应用于输出结果的计算方式，可选的string值有: ``'none'``, ``'mean'``, ``'sum'`` ，对应的int值分别为2，1，0 。默认为 ``'none'``。 即可

Comment (gglin001) [reply python/paddle/fluid/layers/loss.py]: fixed;

去除了
```
This is equivalent to calling
    ``x.backward()`` on a tensor ``x`` when running on the CPU.
```

reduction 部分已应用建议的修改, 麻烦再看下还有没有其他的问题

**Comment by gglin001** [comment]
CI `PR-CI-CINN` 的 fix 见
- https://github.com/PaddlePaddle/Paddle/pull/43860

File: python/paddle/fluid/layers/loss.py [hunk 4]
```diff
@@ -1230,6 +1231,62 @@ def softmax_with_cross_entropy(logits,
...
+    .. math::
+        Out = MEAN(Out)
+
+    When `reduction` is `sum`, return
+
+    .. math::
+        Out = SUM(Out)
+
+    Parameters:
+        x (Variable): The calculated loss ``Tensor``.
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/loss.py]: 这里需要与中文语义保持一致

File: python/paddle/fluid/layers/loss.py (hunk 1 above)
Comment (TCChenlong) [reply python/paddle/fluid/layers/loss.py]: 请删除 This function，文档里不这么描述。

Comment (gglin001) [reply python/paddle/fluid/layers/loss.py]: 抱歉, 之前以为这里是取消这里的 comment, will fix later

File: python/paddle/fluid/layers/loss.py (hunk 4 above)
Comment (czr-gc) [reply python/paddle/fluid/layers/loss.py]: 更正为：
```
        x (Variable): The input tensor. The shapes is [N, *], where N is batch size and `*` means any number of
             additional dimensions. It's data type should be float32, float64 on CPU and float16, float32 on IPU.
```

**Comment by gglin001** [comment]
@TCChenlong hi, 麻烦有空再看下还有其他要修改的地方吗?

关联的 doc 
https://github.com/PaddlePaddle/docs/pull/4932

File: python/paddle/fluid/layers/loss.py [hunk 5]
```diff
@@ -1230,6 +1231,63 @@ def softmax_with_cross_entropy(logits,
...
+
+    When `reduction` is `sum`, return
+
+    .. math::
+        Out = SUM(Out)
+
+    Parameters:
+        x (Variable): The input tensor. The shapes is [N, *], where N is batch size and `*` means any number of
+             additional dimensions. It's data type should be float32, float64 on CPU and float16, float32 on IPU.
+        reduction(str|int): Reduce the loss output. Supported string values are: 'sum', 'mean', 'none'
```
Comment (TCChenlong) [inline python/paddle/fluid/layers/loss.py]: reduction(str|int) ->  reduction(str|int, optional)

Comment (czr-gc) [reply python/paddle/fluid/layers/loss.py]: 已更改为 `reduction(str|int, optional):`

**Review by TCChenlong — DISMISSED** [review]
LGTM

**Review by qili93 — APPROVED** [review]
LGTM

**Review by XiaoguangHu01 — APPROVED** [review]
LGMT

**Review by TCChenlong — APPROVED** [review]
